pkg/apis/core/validation/validation.go

```diff
@@ -3024,10 +3022,70 @@ func validateInitContainers(containers []core.Container, otherContainers []core.
 			allErrs = append(allErrs, field.Invalid(idxPath.Child("startupProbe"), ctr.StartupProbe, "must not be set for init containers"))
 		}
 	}
+
+	return allErrs
+}
+
+func validateContainer(ctr *core.Container, volumes map[string]core.VolumeSource, path *field.Path, opts PodValidationOptions) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	namePath := path.Child("name")
+	if len(ctr.Name) == 0 {
+		allErrs = append(allErrs, field.Required(namePath, ""))
+	} else {
+		allErrs = append(allErrs, ValidateDNS1123Label(ctr.Name, namePath)...)
+	}
+
+	// TODO: do not validate leading and trailing whitespace to preserve backward compatibility.
+	// for example: https://github.com/openshift/origin/issues/14659 image = " " is special token in pod template
+	// others may have done similar
+	if len(ctr.Image) == 0 {
+		allErrs = append(allErrs, field.Required(path.Child("image"), ""))
+	}
+	if ctr.Lifecycle != nil {
+		allErrs = append(allErrs, validateLifecycle(ctr.Lifecycle, path.Child("lifecycle"))...)
+	}
+
+	allErrs = append(allErrs, validateProbe(ctr.LivenessProbe, path.Child("livenessProbe"))...)
+	// Readiness-specific validation
+	if ctr.ReadinessProbe != nil && ctr.ReadinessProbe.TerminationGracePeriodSeconds != nil {
+		allErrs = append(allErrs, field.Invalid(path.Child("readinessProbe", "terminationGracePeriodSeconds"), ctr.ReadinessProbe.TerminationGracePeriodSeconds, "must not be set for readinessProbes"))
+	}
+	allErrs = append(allErrs, validateProbe(ctr.StartupProbe, path.Child("startupProbe"))...)
+	// Liveness-specific validation
+	if ctr.LivenessProbe != nil && ctr.LivenessProbe.SuccessThreshold != 1 {
+		allErrs = append(allErrs, field.Invalid(path.Child("livenessProbe", "successThreshold"), ctr.LivenessProbe.SuccessThreshold, "must be 1"))
+	}
+	allErrs = append(allErrs, validateProbe(ctr.StartupProbe, path.Child("startupProbe"))...)
```

Comment:
Why are we validating startupProbe twice?

Reply:
In fact this whole section is suspect (3049-3063) - the comments on which probe is being validated do not match the code?

Reply:
I agree that is ugly.  Why can't we pass down enough information to skip the "inner" validations when then probe itself is not allowed?